R/recursive_unit_tests.R

```diff
@@ -0,0 +1,192 @@
+RecursiveUnitTests <- R6::R6Class("RecursiveUnitTests",       # nolint
+  public = list(
+    initialize = function(path, filter = "test-.+\\.R$", recursive = TRUE) {
+      if (length(path) > 1) {
+        cli::cli_abort("Please provide a single path.")
+      }
+      private$path <- path
+      private$filter <- filter
+      private$recursive <- recursive
+
+      private$get_valid_test_paths()
+    },
+    run_tests = function(...) {
+      if (private$is_single_test_file()) {
+        testthat::test_file(path = private$valid_paths, ...)
+      } else if (private$is_single_test_dir()) {
+        testthat::test_dir(path = private$valid_paths, ...)
+      } else if (private$is_multiple_test_dirs()) {
+        private$test_recursive(...)
+      } else {
+        cli::cli_abort("Test run failed!")
+      }
+    }
+  ),
+  private = list(
+    filter = NULL,
+    path = NULL,
+    valid_paths = NULL,
+    recursive = TRUE,
+    get_valid_test_paths = function() {
+      if (fs::is_file(private$path)) {
+        private$valid_paths <- fs::path_filter(private$path, regexp = private$filter)
+      }
+
+      if (fs::is_dir(private$path)) {
+        valid_paths <- unique(
+          fs::path_dir(
+            fs::dir_ls(path = private$path,
+                       regexp = private$filter,
+                       recurse = private$recursive, type = "file")
+          )
+        )
+
+        private$valid_paths <- valid_paths[order(valid_paths)]
```

Comment:
Because `testthat::test_dir()` throws an error if given a path without a valid test file, we get list of valid test files, then get their paths.